src/OpenLoco/S5/SawyerStream.cpp

```diff
@@ -0,0 +1,488 @@
+            for (size_t j = 0; j < readLength; j++)
+            {
+                actualChecksum += buffer[j];
+            }
+        }
+
+        valid = checksum == actualChecksum;
+    }
+
+    // Restore position
+    _stream.seekg(backupPos);
+
+    return valid;
+}
+
+void SawyerStreamReader::close()
+{
+    _stream.close();
+}
+
+stdx::span<uint8_t const> SawyerStreamReader::decode(SawyerEncoding encoding, stdx::span<uint8_t const> data)
+{
+    switch (encoding)
+    {
+        case SawyerEncoding::uncompressed:
+            return data;
+        case SawyerEncoding::runLengthSingle:
+            _decodeBuffer2.clear();
+            _decodeBuffer2.reserve(data.size());
+            decodeRunLengthSingle(_decodeBuffer2, data);
+            return _decodeBuffer2.getSpan();
+        case SawyerEncoding::runLengthMulti:
+            _decodeBuffer2.clear();
+            _decodeBuffer2.reserve(data.size());
+            decodeRunLengthSingle(_decodeBuffer2, data);
+
+            _decodeBuffer.clear();
+            _decodeBuffer.reserve(_decodeBuffer2.size());
+            decodeRunLengthMulti(_decodeBuffer, _decodeBuffer2.getSpan());
+            return _decodeBuffer.getSpan();
+        case SawyerEncoding::rotate:
+            _decodeBuffer2.clear();
+            _decodeBuffer2.reserve(data.size());
+            decodeRotate(_decodeBuffer2, data);
+            return _decodeBuffer2.getSpan();
+        default:
+            throw std::runtime_error(exceptionUnknownEncoding);
+    }
+}
+
+void SawyerStreamReader::decodeRunLengthSingle(FastBuffer& buffer, stdx::span<uint8_t const> data)
+{
+    for (size_t i = 0; i < data.size(); i++)
+    {
+        uint8_t rleCodeByte = data[i];
+        if (rleCodeByte & 128)
+        {
+            i++;
+            if (i >= data.size())
+            {
+                throw std::runtime_error(exceptionInvalidRLE);
+            }
+
+            auto copyLen = static_cast<size_t>(257 - rleCodeByte);
+            auto copyByte = data[i];
+            buffer.push_back(copyByte, copyLen);
+        }
+        else
+        {
+            if (i + 1 >= data.size() || i + 1 + rleCodeByte + 1 > data.size())
+            {
+                throw std::runtime_error(exceptionInvalidRLE);
+            }
+
+            auto copyLen = static_cast<size_t>(rleCodeByte + 1);
+            buffer.push_back(&data[i + 1], copyLen);
+            i += rleCodeByte + 1;
+        }
+    }
+}
+
+void SawyerStreamReader::decodeRunLengthMulti(FastBuffer& buffer, stdx::span<uint8_t const> data)
+{
+    for (size_t i = 0; i < data.size(); i++)
+    {
+        if (data[i] == 0xFF)
+        {
+            i++;
+            if (i >= data.size())
+            {
+                throw std::runtime_error(exceptionInvalidRLE);
+            }
+            buffer.push_back(data[i]);
+        }
+        else
+        {
+            auto offset = static_cast<int32_t>(data[i] >> 3) - 32;
```

Comment:
do we need this cast. Seems pointless

Reply:
Because it needs to convert to a signed number so that it can be negative.